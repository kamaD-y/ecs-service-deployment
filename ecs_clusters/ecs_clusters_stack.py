import glob
from pathlib import Path
from typing import List, Optional, TypedDict

import yaml
from aws_cdk import CfnOutput, Stack
from aws_cdk import aws_ec2 as ec2
from aws_cdk import aws_ecs as ecs
from constructs import Construct


class ClusterService(TypedDict):
    service_name: str
    task_arn: str
    desired_count: int
    task_definition: ecs.ITaskDefinition


class ClusterInfo(TypedDict):
    name: str
    cluster: ecs.ICluster
    services: List[ClusterService]


# 既存リソースを取得するコンストラクタ
class ExistingResourceConstructor(Construct):
    def __init__(
        self,
        scope: Construct,
        id: str,
        vpc_id: str,
        sg_ids: List[str],
        cluster_services_glob_path: str,
    ) -> None:
        super().__init__(scope, id)
        self.vpc: Optional[ec2.IVpc] = None
        self.security_groups: List[ec2.ISecurityGroup] = []
        self.clusters: List[ClusterInfo] = []
        self.load_existing_resources(vpc_id, sg_ids, cluster_services_glob_path)

    def _load_existing_vpc(self, vpc_id: str) -> None:
        self.vpc = ec2.Vpc.from_lookup(self, "ImportedVpc", vpc_id=vpc_id)

    def _load_existing_security_groups(self, sg_ids: list) -> None:
        for i, sg_id in enumerate(sg_ids):
            self.security_groups.append(
                ec2.SecurityGroup.from_security_group_id(self, f"ImportedSecurityGroup{i}", security_group_id=sg_id)
            )

    def _load_existing_clusters(self, cluster_services_glob_path: str) -> None:
        for file_path in glob.glob(cluster_services_glob_path):
            cluster_info = {}
            cluster_name = Path(file_path).stem
            cluster_info["name"] = cluster_name
            cluster_info["cluster"] = ecs.Cluster.from_cluster_attributes(
                self,
                f"ImportedCluster-{cluster_name}",
                cluster_name=cluster_name,
                vpc=self.vpc,
                security_groups=self.security_groups,
            )
            with open(file_path, "r") as file:
                cluster_info["services"] = yaml.safe_load(file)
            for service in cluster_info["services"]:
                service["task_definition"] = ecs.TaskDefinition.from_task_definition_arn(
                    self,
                    f"ImportedTaskDef-{cluster_name}-{service['service_name']}",
                    task_definition_arn=service["task_arn"],
                )
            self.clusters.append(cluster_info)

    def load_existing_resources(self, vpc_id: str, sg_ids: list, cluster_services_glob_path: str) -> None:
        self._load_existing_vpc(vpc_id)
        self._load_existing_security_groups(sg_ids)
        self._load_existing_clusters(cluster_services_glob_path)
        # print(self.clusters)
        """
        [{
          'name': 'fargate-cluster-1',
          'cluster': <jsii._reference_map.InterfaceDynamicProxy object at 0x7f4618d7a000>,
          'services': [
            {'name': 'fargate-service-1',
            'task_arn': 'arn:aws:ecs:ap-northeast-1:123456789012:task-definition/task-1:1',
            'desired_count': 2,
            'task_definition': <jsii._reference_map.InterfaceDynamicProxy object at 0x7f4618cf5dc0>
            },
            {'name': 'fargate-service-2',
            'task_arn': 'arn:aws:ecs:ap-northeast-1:123456789012:task-definition/task-1:2',
            'desired_count': 3,
            'task_definition': <jsii._reference_map.InterfaceDynamicProxy object at 0x7f4618cf5760>
            },
          ]
        }]
        """


class EcsClustersStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, parameters, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        """
        既存リソース参照
        """
        vpc_id = parameters["vpc_id"]
        sg_ids = parameters["sg_ids"]
        cluster_services_glob_path = parameters["cluster_services_glob_path"]

        existing_resource = ExistingResourceConstructor(
            self, "ExistingResources", vpc_id, sg_ids, cluster_services_glob_path
        )

        """
        サービス作成
        """
        # 必須の何かがITaskDefinitionでは足りないらしく、既存のタスク定義では不可の為仕方なく一旦TaskDefinitionを作成する
        # 一時的にサービスにデフォルトタスク定義を紐づけておき、後で既存のタスク定義を紐づける
        # ref: https://github.com/aws/aws-cdk/issues/7863
        default_task_definition = ecs.TaskDefinition(
            self,
            "DefaultTaskDef",
            family="DefaultTaskDef",
            compatibility=ecs.Compatibility.FARGATE,
            cpu="256",
            memory_mib="512",
            network_mode=ecs.NetworkMode.AWS_VPC,
            task_role=None,
            execution_role=None,
        )
        default_task_definition.add_container(
            "DefaultContainer",
            image=ecs.ContainerImage.from_registry("amazon/amazon-ecs-sample"),
            memory_limit_mib=512,
            cpu=256,
        )

        # 各クラスターのサービスを作成
        for cluster in existing_resource.clusters:
            for cluster_service in cluster["services"]:
                service_name = cluster_service["service_name"]

                # Fargateサービスを作成, 一時的にデフォルトタスク定義を関連付ける
                fargate_service = ecs.FargateService(
                    self,
                    f"FargateService-{cluster['name']}-{service_name}",
                    cluster=cluster["cluster"],
                    task_definition=default_task_definition,
                    desired_count=cluster_service["desired_count"],
                    service_name=service_name,
                    assign_public_ip=True,  # 必要に応じて設定
                )
                # 既存のタスク定義を関連付けし直す
                fargate_service.node.try_find_child("Service").task_definition = cluster_service[
                    "task_definition"
                ].task_definition_arn

                CfnOutput(
                    self,
                    f"ServiceARN-{cluster['name']}-{service_name}",
                    value=fargate_service.service_arn,
                    description=f"ARN of the {service_name} service",
                )
