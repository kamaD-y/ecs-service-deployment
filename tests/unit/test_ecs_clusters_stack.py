import aws_cdk as cdk
import aws_cdk.assertions as assertions

from ecs_clusters.ecs_clusters_stack import EcsClustersStack


def test_snapshot(snapshot):
    app = cdk.App()
    stack = EcsClustersStack(
        app,
        "ecs-clusters",
        parameters={
            "vpc_id": "vpc-12345678",
            "sg_ids": ["sg-12345678"],
            "cluster_services_glob_path": "config/cluster-services/*.yml",
        },
        env=cdk.Environment(account="123456789012", region="ap-northeast-1"),
    )
    template = assertions.Template.from_stack(stack)

    assert template.to_json() == snapshot
