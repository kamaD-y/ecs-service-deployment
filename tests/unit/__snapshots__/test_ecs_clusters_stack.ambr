# serializer version: 1
# name: test_snapshot
  dict({
    'Outputs': dict({
      'ServiceARNfargatecluster1fargateservice1': dict({
        'Description': 'ARN of the fargate-service-1 service',
        'Value': dict({
          'Ref': 'FargateServicefargatecluster1fargateservice1ServiceA2DA3509',
        }),
      }),
      'ServiceARNfargatecluster2fargateservice1': dict({
        'Description': 'ARN of the fargate-service-1 service',
        'Value': dict({
          'Ref': 'FargateServicefargatecluster2fargateservice1ServiceA37EF0FF',
        }),
      }),
    }),
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'DefaultTaskDefF829348D': dict({
        'Properties': dict({
          'ContainerDefinitions': list([
            dict({
              'Cpu': 256,
              'Essential': True,
              'Image': 'amazon/amazon-ecs-sample',
              'Memory': 512,
              'Name': 'DefaultContainer',
            }),
          ]),
          'Cpu': '256',
          'Family': 'DefaultTaskDef',
          'Memory': '512',
          'NetworkMode': 'awsvpc',
          'RequiresCompatibilities': list([
            'FARGATE',
          ]),
          'TaskRoleArn': dict({
            'Fn::GetAtt': list([
              'DefaultTaskDefTaskRoleD38D9022',
              'Arn',
            ]),
          }),
        }),
        'Type': 'AWS::ECS::TaskDefinition',
      }),
      'DefaultTaskDefTaskRoleD38D9022': dict({
        'Properties': dict({
          'AssumeRolePolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'sts:AssumeRole',
                'Effect': 'Allow',
                'Principal': dict({
                  'Service': 'ecs-tasks.amazonaws.com',
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
        }),
        'Type': 'AWS::IAM::Role',
      }),
      'FargateServicefargatecluster1fargateservice1SecurityGroup0786FAFE': dict({
        'DependsOn': list([
          'DefaultTaskDefTaskRoleD38D9022',
        ]),
        'Properties': dict({
          'GroupDescription': 'ecs-clusters/FargateService-fargate-cluster-1-fargate-service-1/SecurityGroup',
          'SecurityGroupEgress': list([
            dict({
              'CidrIp': '0.0.0.0/0',
              'Description': 'Allow all outbound traffic by default',
              'IpProtocol': '-1',
            }),
          ]),
          'VpcId': 'vpc-12345',
        }),
        'Type': 'AWS::EC2::SecurityGroup',
      }),
      'FargateServicefargatecluster1fargateservice1ServiceA2DA3509': dict({
        'DependsOn': list([
          'DefaultTaskDefTaskRoleD38D9022',
        ]),
        'Properties': dict({
          'Cluster': 'fargate-cluster-1',
          'DeploymentConfiguration': dict({
            'Alarms': dict({
              'AlarmNames': list([
              ]),
              'Enable': False,
              'Rollback': False,
            }),
            'MaximumPercent': 200,
            'MinimumHealthyPercent': 50,
          }),
          'DesiredCount': 1,
          'EnableECSManagedTags': False,
          'LaunchType': 'FARGATE',
          'NetworkConfiguration': dict({
            'AwsvpcConfiguration': dict({
              'AssignPublicIp': 'ENABLED',
              'SecurityGroups': list([
                dict({
                  'Fn::GetAtt': list([
                    'FargateServicefargatecluster1fargateservice1SecurityGroup0786FAFE',
                    'GroupId',
                  ]),
                }),
              ]),
              'Subnets': list([
                's-12345',
                's-67890',
              ]),
            }),
          }),
          'ServiceName': 'fargate-service-1',
          'TaskDefinition': 'arn:aws:ecs:ap-northeast-1:123456789012:task-definition/task-1',
        }),
        'Type': 'AWS::ECS::Service',
      }),
      'FargateServicefargatecluster2fargateservice1SecurityGroupB6718C5A': dict({
        'DependsOn': list([
          'DefaultTaskDefTaskRoleD38D9022',
        ]),
        'Properties': dict({
          'GroupDescription': 'ecs-clusters/FargateService-fargate-cluster-2-fargate-service-1/SecurityGroup',
          'SecurityGroupEgress': list([
            dict({
              'CidrIp': '0.0.0.0/0',
              'Description': 'Allow all outbound traffic by default',
              'IpProtocol': '-1',
            }),
          ]),
          'VpcId': 'vpc-12345',
        }),
        'Type': 'AWS::EC2::SecurityGroup',
      }),
      'FargateServicefargatecluster2fargateservice1ServiceA37EF0FF': dict({
        'DependsOn': list([
          'DefaultTaskDefTaskRoleD38D9022',
        ]),
        'Properties': dict({
          'Cluster': 'fargate-cluster-2',
          'DeploymentConfiguration': dict({
            'Alarms': dict({
              'AlarmNames': list([
              ]),
              'Enable': False,
              'Rollback': False,
            }),
            'MaximumPercent': 200,
            'MinimumHealthyPercent': 50,
          }),
          'DesiredCount': 1,
          'EnableECSManagedTags': False,
          'LaunchType': 'FARGATE',
          'NetworkConfiguration': dict({
            'AwsvpcConfiguration': dict({
              'AssignPublicIp': 'ENABLED',
              'SecurityGroups': list([
                dict({
                  'Fn::GetAtt': list([
                    'FargateServicefargatecluster2fargateservice1SecurityGroupB6718C5A',
                    'GroupId',
                  ]),
                }),
              ]),
              'Subnets': list([
                's-12345',
                's-67890',
              ]),
            }),
          }),
          'ServiceName': 'fargate-service-1',
          'TaskDefinition': 'arn:aws:ecs:ap-northeast-1:123456789012:task-definition/task-1:3',
        }),
        'Type': 'AWS::ECS::Service',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
